source/extensions/http/header_validators/envoy_default/http1_header_validator.cc

```diff
@@ -0,0 +1,368 @@
+  const std::string TransferEncodingNotAllowed = "uhv.http1.transfer_encoding_not_allowed";
+  const std::string ContentLengthNotAllowed = "uhv.http1.content_length_not_allowed";
+  const std::string ChunkedContentLength = "uhv.http1.content_length_and_chunked_not_allowed";
+};
+
+using Http1ResponseCodeDetail = ConstSingleton<Http1ResponseCodeDetailValues>;
+
+/*
+ * Header validation implementation for the Http/1 codec. This class follows guidance from
+ * several RFCS:
+ *
+ * RFC 3986 <https://datatracker.ietf.org/doc/html/rfc3986> URI Generic Syntax
+ * RFC 7230 <https://datatracker.ietf.org/doc/html/rfc7230> HTTP/1.1 Message Syntax
+ * RFC 7231 <https://datatracker.ietf.org/doc/html/rfc7231> HTTP/1.1 Semantics and Content
+ *
+ */
+Http1HeaderValidator::Http1HeaderValidator(const HeaderValidatorConfig& config, Protocol protocol,
+                                           StreamInfo::StreamInfo& stream_info)
+    : HeaderValidator(config, protocol, stream_info) {}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http1HeaderValidator::validateRequestHeaderEntry(const HeaderString& key,
+                                                 const HeaderString& value) {
+  static const absl::node_hash_map<absl::string_view, HeaderValidatorFunction> kHeaderValidatorMap{
+      {":method", &Http1HeaderValidator::validateMethodHeader},
+      {":authority", &Http1HeaderValidator::validateHostHeader},
+      {"host", &Http1HeaderValidator::validateHostHeader},
+      {":scheme", &Http1HeaderValidator::validateSchemeHeader},
+      {":path", &Http1HeaderValidator::validateGenericPathHeader},
+      {"transfer-encoding", &Http1HeaderValidator::validateTransferEncodingHeader},
+      {"content-length", &Http1HeaderValidator::validateContentLengthHeader},
+  };
+
+  const auto& key_string_view = key.getStringView();
+  if (key_string_view.empty()) {
+    // reject empty header names
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().EmptyHeaderName};
+  }
+
+  auto validator_it = kHeaderValidatorMap.find(key_string_view);
+  if (validator_it != kHeaderValidatorMap.end()) {
+    const auto& validator = validator_it->second;
+    return (*this.*validator)(value);
+  }
+
+  if (key_string_view.at(0) != ':') {
+    // Validate the (non-pseudo) header name
+    auto name_result = validateGenericHeaderName(key);
+    if (!name_result) {
+      return name_result;
+    }
+  } else {
+    // kHeaderValidatorMap contains every known pseudo header. If the header name starts with ":"
+    // and we don't have a validator registered in the map, then the header name is an unknown
+    // pseudo header.
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidPseudoHeader};
+  }
+
+  return validateGenericHeaderValue(value);
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http1HeaderValidator::validateResponseHeaderEntry(const HeaderString& key,
```

Comment:
cc @bencebeky can you take a look at the HTTP/1.1 validation with an eye to where the defaults might be problematic for Google?

Reply:
Nothing sticks out to me as potentially problematic.